src/psyclone/parse/algorithm.py

```diff
@@ -753,19 +774,42 @@ def get_kernel(parse_tree, alg_filename, arg_type_defns):
             if not walk(argument, Name):
                 # This is a literal so store the full expression as a
                 # string
-                arguments.append(Arg('literal', argument.tostr().lower()))
+                # First determine the datatype
+                candidate_datatype = None
+                for literal in walk(
+                        argument, (Real_Literal_Constant,
+                                   Int_Literal_Constant)):
+                    precision = literal.children[1]
+                    if precision:
+                        precision = str(precision)
+                    if isinstance(literal, Real_Literal_Constant):
+                        datatype = ("real", precision)
+                    else:  # it's an Int_Literal_Constant
+                        datatype = ("integer", precision)
+
+                    if not candidate_datatype:
+                        # This is the first candidate
+                        candidate_datatype = datatype
+                    elif candidate_datatype != datatype:
+                        raise NotImplementedError(
+                            f"Found two non-matching literals within an "
+                            f"expression passed into an invoke from the "
+                            f"algorithm layer. '{datatype}' and "
+                            f"'{candidate_datatype}' do not match. "
+                            f"This is not supported in PSyclone.")
+                arguments.append(Arg('literal', argument.tostr().lower(),
+                                 datatype=datatype))
             else:
                 raise NotImplementedError(
```

Comment:
You could do `if walk(argument, Name): then raise NotImplementedError` first and then you could reduce the indentation of the code above.

Reply:
Good point.